libtock/services/alarm.c

```diff
@@ -75,19 +70,44 @@ static uint32_t ticks_to_ms(uint32_t ticks) {
 static libtock_alarm_ticks_t* root = NULL;
 
 static void root_insert(libtock_alarm_ticks_t* alarm) {
+  // We want to insert the new alarm just before an existing alarm
+  // that should expire next after it, as `alarm_upcall` breaks as
+  // soon as it finds an alarm that should not fire yet. To do so, we
+  // need to account for an empty list, a non-empty list where the new
+  // alarm should be last, as well as computing relative expirations
+  // when alarm expirations may overflow the clock 0 or 1 times.
+
   if (root == NULL) {
     root       = alarm;
     root->next = NULL;
     root->prev = NULL;
     return;
   }
 
+  // Compute the tick value at which the new alarm will expire.
+  uint32_t new_expiration = alarm->reference + alarm->dt;
+  // Determine whether the clock overflows before the new alarm
+  // expires. Because ticks are 32-bit, a clock can overflow at most
+  // once before a 32-bit alarm fires.
+  bool new_overflows = alarm->reference > UINT32_MAX - alarm->dt;
+
   libtock_alarm_ticks_t** cur = &root;
   libtock_alarm_ticks_t* prev = NULL;
   while (*cur != NULL) {
+    // Compute the tick value at which this alarm will expire.
     uint32_t cur_expiration = (*cur)->reference + (*cur)->dt;
-    uint32_t new_expiration = alarm->reference + alarm->dt;
-    if (!within_range(alarm->reference, cur_expiration, new_expiration)) {
+    // Determine whether the clock overflows before this alarm
+    // expires.
+    bool cur_overflows = (*cur)->reference > UINT32_MAX - (*cur)->dt;
+
+    // This alarm happens after the new alarm if:
```

Comment:
What is "this alarm" and what is "the new alarm"? I think it's `cur` and `alarm` respectively? But I might have that backwards. In my opinion, this entire comment block should use the actual names for clarity.